syntax = "proto3";

package ipvc.integration;

// Dataset Service for data integration operations
service DatasetService {
  // List all datasets
  rpc ListDatasets(ListDatasetsRequest) returns (ListDatasetsResponse);
  
  // Get dataset details
  rpc GetDataset(GetDatasetRequest) returns (Dataset);
  
  // Get dataset records
  rpc GetDatasetRecords(GetDatasetRecordsRequest) returns (GetDatasetRecordsResponse);
  
  // Generate XML for dataset
  rpc GenerateXML(GenerateXMLRequest) returns (GenerateXMLResponse);
  
  // Validate XML
  rpc ValidateXML(ValidateXMLRequest) returns (ValidateXMLResponse);
  
  // Stream dataset records
  rpc StreamDatasetRecords(StreamRecordsRequest) returns (stream DataRecord);
}

// Request/Response Messages

message ListDatasetsRequest {
  int32 page = 1;
  int32 page_size = 2;
}

message ListDatasetsResponse {
  repeated DatasetSummary datasets = 1;
  int32 total_count = 2;
  int32 page = 3;
  int32 page_size = 4;
}

message GetDatasetRequest {
  int32 id = 1;
}

message GetDatasetRecordsRequest {
  int32 dataset_id = 1;
  int32 limit = 2;
  int32 offset = 3;
}

message GetDatasetRecordsResponse {
  repeated DataRecord records = 1;
  int32 total_rows = 2;
  int32 offset = 3;
  int32 limit = 4;
}

message GenerateXMLRequest {
  int32 dataset_id = 1;
  int32 limit = 2;
  bool generate_xsd = 3;
  bool validate = 4;
}

message GenerateXMLResponse {
  string dataset_name = 1;
  bool xsd_generated = 2;
  string xsd_path = 3;
  bool xml_generated = 4;
  string xml_path = 5;
  bool validation_passed = 6;
  repeated string validation_errors = 7;
}

message ValidateXMLRequest {
  int32 dataset_id = 1;
}

message ValidateXMLResponse {
  string dataset_name = 1;
  bool is_valid = 2;
  repeated string errors = 3;
}

message StreamRecordsRequest {
  int32 dataset_id = 1;
  int32 batch_size = 2;
}

// Data Models

message DatasetSummary {
  int32 id = 1;
  string name = 2;
  string description = 3;
  int32 total_rows = 4;
  int32 total_columns = 5;
  string status = 6;
  string created_at = 7;
  string updated_at = 8;
}

message Dataset {
  int32 id = 1;
  string name = 2;
  string description = 3;
  string source_file = 4;
  int32 total_rows = 5;
  int32 total_columns = 6;
  string status = 7;
  string created_at = 8;
  string updated_at = 9;
  string imported_at = 10;
  string xml_file_path = 11;
  string xml_schema_path = 12;
  repeated DatasetColumn columns = 13;
}

message DatasetColumn {
  int32 id = 1;
  string name = 2;
  string data_type = 3;
  bool nullable = 4;
  bool unique = 5;
  bool primary_key = 6;
  int32 null_count = 7;
  int32 unique_count = 8;
  int32 position = 9;
  repeated string sample_values = 10;
}

message DataRecord {
  int32 id = 1;
  int32 dataset_id = 2;
  int32 row_number = 3;
  string data_json = 4;  // JSON string representation of data
  string created_at = 5;
}
